src/request.rs

```diff
@@ -90,23 +88,30 @@ pub struct Request {
     pub hostname: String,
     pub source_hostname_hashes: Option<Vec<utils::Hash>>,
 
+    pub(crate) url_lower_cased: String,
+    pub(crate) request_tokens: Vec<utils::Hash>,
     pub(crate) original_url: String,
 }
 
 impl Request {
-    pub(crate) fn get_url(&self, case_sensitive: bool) -> std::borrow::Cow<str> {
+    pub(crate) fn get_url(&self, case_sensitive: bool) -> &str {
         if case_sensitive {
-            Cow::Borrowed(&self.url)
+            &self.url
         } else {
-            Cow::Owned(self.url.to_ascii_lowercase())
+            &self.url_lower_cased
         }
     }
 
-    pub fn get_tokens(&self, token_buffer: &mut Vec<utils::Hash>) {
-        token_buffer.clear();
-        utils::tokenize_pooled(&self.url.to_ascii_lowercase(), token_buffer);
-        // Add zero token as a fallback to wildcard rule bucket
-        token_buffer.push(0);
+    pub fn get_tokens_for_match(&self) -> impl Iterator<Item = &utils::Hash> {
+        self.source_hostname_hashes
```

Comment:
Let's add a small comment that `source_hostname_hashes` should be the first for performance reasons.